NineChronicles.Headless/GraphTypes/StandaloneQuery.cs

```diff
@@ -72,6 +74,87 @@ public StandaloneQuery(StandaloneContext standaloneContext, IConfiguration confi
                 }
             );
 
+            Field<ListGraphType<RootStateDiffType>>(
+                name: "diffs",
+                arguments: new QueryArguments(
+                    new QueryArgument<NonNullGraphType<LongGraphType>>
+                    {
+                        Name = "baseIndex",
+                        Description = "The index of the block used to fetch state from chain."
+                    },
+                    new QueryArgument<NonNullGraphType<LongGraphType>>
+                    {
+                        Name = "changedIndex",
+                        Description = "The index of the block used to fetch state from chain."
+                    }
+                ),
+                resolve: context =>
+                {
+                    if (!(standaloneContext.BlockChain is BlockChain blockChain))
+                    {
+                        throw new ExecutionError(
+                            $"{nameof(StandaloneContext)}.{nameof(StandaloneContext.BlockChain)} was not set yet!"
+                        );
+                    }
+
+                    var baseIndex = context.GetArgument<long>("baseIndex");
+                    var changedIndex = context.GetArgument<long>("changedIndex");
+
+                    var blockInterval = changedIndex - baseIndex;
+
+                    if (blockInterval >= 10 && blockInterval > 0)
+                    {
+                        throw new ExecutionError(
+                            "Interval between baseIndex and changedIndex should under 10"
+                        );
+                    }
+
+                    var baseBlockStateRootHash = blockChain[baseIndex].StateRootHash.ToString();
+                    var changedBlockStateRootHash = blockChain[changedIndex].StateRootHash.ToString();
+
+                    var baseStateRootHash = HashDigest<SHA256>.FromString(baseBlockStateRootHash);
+                    var targetStateRootHash = HashDigest<SHA256>.FromString(
+                        changedBlockStateRootHash
+                    );
+
+                    var stateStore = standaloneContext.StateStore;
+                    var baseTrieModel = stateStore.GetStateRoot(baseStateRootHash);
+                    var targetTrieModel = stateStore.GetStateRoot(targetStateRootHash);
+
+                    var diffs = baseTrieModel
+                        .Diff(targetTrieModel)
+                        .Select(x =>
+                        {
+                            var baseSubTrieModel = stateStore.GetStateRoot(
+                                new HashDigest<SHA256>((Binary)x.SourceValue)
+                            );
+                            if (x.TargetValue is not null)
```

Comment:
We should handle the case when `TargetValue` is `null`, too. Maybe we should resolve issue with the *union* type.

In TypeScript:

```
// If TargetValue is not null, it can be root of subtrie. => { path: string, diffs: StateDiffType[] }
// If TargetValue is null, it means the subtrie has removed. => StateDiffType.
type RootStateDiffType = { path: string; diffs: StateDiffType[] } | StateDiffType;
type StateDiffType = { path: string; baseState: unknown; changedState: unknown };
```

Reply:
It's meant to be like this, right?

```
var diffs = baseTrieModel
    .Diff(targetTrieModel)
// Type error
    .Select(x =>
    {
        if (x.TargetValue is not null)
        {
            var targetSubTrieModel = stateStore.GetStateRoot(new HashDigest<SHA256>((Binary)x.TargetValue));
            return new RootStateDiffType.Value(
                Encoding.Default.GetString(x.Path.ByteArray.ToArray()),
                targetSubTrieModel.Diff(baseTrieModel)
                    .Select(diff => new StateDiffType.Value(
                        Encoding.Default.GetString(diff.Path.ByteArray.ToArray()),
                        diff.SourceValue,
                        diff.TargetValue
                    )).ToArray()
            );
        }
        else
        {
            return new StateDiffType.Value(
                Encoding.Default.GetString(x.Path.ByteArray.ToArray()),
                x.SourceValue,
                x.TargetValue
            );
        }
    }).ToArray();
```

Reply:
From what I understand, it's tough to use union types directly in .NET (thankfully, GraphQL has them).
So, for the 'diffs' variable, it seems like we need to use a sort of fake, upper-level interface and handle it that way. Does that sound right?

```csharp
var diffs = baseTrieModel
    .Diff(targetTrieModel)
    .Select(x =>
    {
        if (x.TargetValue is not null)
        {
            var targetSubTrieModel = stateStore.GetStateRoot(new HashDigest<SHA256>((Binary)x.TargetValue));
            return (IDiffResult)new RootStateDiffTypeValue(
                Encoding.Default.GetString(x.Path.ByteArray.ToArray()),
                targetSubTrieModel.Diff(baseTrieModel)
                    .Select(diff => new StateDiffTypeValue(
                        Encoding.Default.GetString(diff.Path.ByteArray.ToArray()),
                        diff.SourceValue,
                        diff.TargetValue
                    )).ToArray()
            );
        }
        else
        {
            return (IDiffResult)new StateDiffTypeValue(
                Encoding.Default.GetString(x.Path.ByteArray.ToArray()),
                x.SourceValue,
                x.TargetValue
            );
        }
    }).ToArray();
```

Reply:
Yes, you can see https://graphql-dotnet.github.io/docs/getting-started/unions/

Reply:
Done!!!!!!!